kong/tools/wrpc/future.lua

```diff
@@ -0,0 +1,85 @@
+local semaphore = require "ngx.semaphore"
+local semaphore_new = semaphore.new
+
+local ngx_log = ngx.log
+local ERR = ngx.ERR
+local ngx_now = ngx.now
+local new_timer = ngx.timer.at
+local setmetatable = setmetatable
+
+local _M = {}
+local _MT = { __index = _M, }
+
+local function finish_future_life(future)
+  future.response_t[future.seq] = nil
+end
+
+local function drop_aftermath(premature, future)
+  if premature then
+    return
+  end
+
+  local ok, err = future:wait()
+  if not ok then
+    ngx_log(ERR, "request fail to recieve response: ", err)
+  end
+end
+
+-- Call to intentionally drop the future, without blocking to wait.
+-- It will discard the response and log if error occurs.
+function _M:drop()
+  return new_timer(0, drop_aftermath, self)
+end
+
+-- Call to indicate the request is done.
+--- @param any data the response
+function _M:done(data)
+  self.data = data
+  self.smph:post()
+  finish_future_life(self)
+end
+
+-- Call to indicate the request is in error.
+--- @param string etype error type enumerator
+--- @param string errdesc the error description
+function _M:error(etype, errdesc)
+  self.data = nil
+  self.etype = etype
+  self.errdesc = errdesc
+  self.smph:post()
+  finish_future_life(self)
+end
+
+-- call to indicate the request expires.
+function _M:expire()
+  self:error("timeout", "timeout")
+end
+
+-- wait until the request is done or in error
+--- @return any data, string err
+function _M:wait()
+  local ok, err = self.smph:wait(self.delay)
+  if not ok then
+    return nil, err
+  end
+
+  return self.data, self.errdesc
+end
+
+--- @param delay number time until deadline
+--- @param wrpc_peer table wRPC peer that creates the call
+function _M.new(wrpc_peer, delay)
```

Comment:
perhaps I am missing some context here. I am having a hard time understanding what a "future" is supposed to represent here, and I am also confused about what the methods do.

Is this similar to the concept of Promise in Node.js, or somewhere else? If this is based on something, I suggest copying the method names from there.

Reply:
i've seen promises and futures used mostly interchangeably.  i'm not sure which i'd prefer.  I thought SICP used "futures", but reviewing it it mostly calls them "expressions with lazy evaluation", and only informally mentions that they represent a "promise".

according to [wikipedia](https://en.wikipedia.org/wiki/Futures_and_promises):

> The terms future, promise, delay, and deferred are often used interchangeably, although some differences in usage between future and promise are treated below. Specifically, when usage is distinguished, a future is a read-only placeholder view of a variable, while a promise is a writable, [single assignment](https://en.wikipedia.org/wiki/Single_assignment) container which sets the value of the future.

in C++11 there are both `promise` and `future` objects, [where](https://www.cplusplus.com/reference/future/promise/):

> - The promise object is the asynchronous provider and is expected to set a value for the shared state at some point.
> - The future object is an asynchronous return object that can retrieve the value of the shared state, waiting for it to be ready, if necessary.

so, the producer constructs a `promise` and from it a `future`.  the "user" gets the `future` and the background process writes to the `promise`, making the `future` ready (an waking anybody that might've been waiting for it)

so.... i'd be (very) slightly inclined to call this a future.